Yesod/Auth/OAuth2.hs

```diff
@@ -102,15 +102,18 @@ authOAuth2Widget widget name oauth getCreds = AuthPlugin name dispatch login
         deleteSession tokenSessionKey
         case newToken of
             Just csrfToken | newToken == oldToken -> do
-                code <- lift $ runInputGet $ ireq textField "code"
-                oauth' <- withCallback csrfToken
-                master <- lift getYesod
-                result <- liftIO $ fetchAccessToken (authHttpManager master) oauth' (encodeUtf8 code)
-                case result of
-                    Left _ -> permissionDenied "Unable to retreive OAuth2 token"
-                    Right token -> do
-                        creds <- liftIO $ getCreds (authHttpManager master) token
-                        lift $ setCredsRedirect creds
+                mcode <- lift $ runInputGet $ iopt textField "code"
```

Comment:
This code's getting a little noisy with all the nested `case`s.

What do you think about wrapping `lookupGetParam` into a `requireGetParam` to cetralize the `Maybe` handling out of the main logic, then relying on `permissionDenied`'s short-circuit nature to bail early:

```hs
dispatch "GET" ["callback"] = do
    code <- requireGetParam "code"
    csrfToken <- requireGetParam "state"

    unless (csrfToken == oldToken) $ permissionDenied "Invalid OAuth2 state token"

    -- main logic...

  where
    requireGetParam key =
        maybe (permissionDenied $ key <> " parameter not provided") return <$> lookupGetParam key
```

Note: not test-compiled